test-data/unit/check-based-intersection.test

```diff
@@ -0,0 +1,77 @@
+[case testIntersection]
+from basedtyping import Intersection
+i: Intersection[int, str]
+reveal_type(i)  # N: Revealed type is "int & str"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionSyntax]
+from __future__ import annotations
+i: int & str
+reveal_type(i)  # N: Revealed type is "int & str"
+[builtins fixtures/tuple.pyi]
+
+
+[case testNoIntersectionSyntax]
+i: int & str  # E: X & Y syntax for intersections requires quoted types or __future__.annotations  [valid-type]
+
+
+[case testIntersectionAttribute]
+from __future__ import annotations
+class A:
+    a: int
+
+class B:
+    b: str
+
+i: A & B
+reveal_type(i.a)  # N: Revealed type is "int"
+reveal_type(i.b)  # N: Revealed type is "str"
+reveal_type(i.c)  # E: "A & B" has no attribute "c"  [attr-defined] \
+                  # N: Revealed type is "Any (from error)"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionAssignability]
+# flags: --disallow-redefinition
+from __future__ import annotations
+class A: pass
+class B: pass
+class C(A, B): pass
+class D: pass
+
+c: C
+cd: C & D
+c_or_d: C | D
+d: D
+i: A & B
+ab: A & B = i
+i = c
+i = d  # E: Incompatible types in assignment (expression has type "D", variable has type "A & B")  [assignment]
+i = c_or_d  # E: Incompatible types in assignment (expression has type "C | D", variable has type "A & B")  [assignment]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionFromNarrowing]
+from __future__ import annotations
+class A: pass
+class B: pass
+
+i: A
+assert isinstance(i, B)
+reveal_type(i)  # N: Revealed type is "__main__.A & __main__.B"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionWithCallable]
+from __future__ import annotations
+from typing import Callable
+class A:
+    def __call__(self, a: int): ...
+foo: A & Callable[[], int]
+reveal_type(foo())  # N: Revealed type is "int"
+foo("")  # E: No overload variant of "A & () -> int" matches argument type "str"  [call-overload] \
+         # N: Possible overload variants: \
+         # N:     def __call__(self, a: int) -> None \
+         # N:     def () -> int
+[builtins fixtures/tuple.pyi]
```

Comment:
what about this but instead `A` has no `__call__`

Reply:
```
test.py:7:1: error: No overload variant of "A & () -> int" matches argument type "str"  [call-overload]
test.py:7:1: note: Possible overload variant:
test.py:7:1: note:     def A() -> int
```
lmao

Reply:
as long as u can still call it with no args

Reply:
```
[case testIntersectionOneCallable]
from __future__ import annotations
from typing import Callable
class A:
    pass
foo: A & Callable[[], int]
reveal_type(foo())  # N: Revealed type is "int"
foo("")  # E: No overload variant of "A & () -> int" matches argument type "str"  [call-overload] \
         # N: Possible overload variant: \
         # N:     def () -> int
[builtins fixtures/tuple.pyi]
```